src/sqlfluff/utils/reflow/classes.py

```diff
@@ -0,0 +1,283 @@
+"""Dataclasses for reflow work."""
+
+
+from itertools import chain
+import logging
+from dataclasses import dataclass
+from typing import List, Sequence, Union, Type
+
+from sqlfluff.core.parser import BaseSegment, RawSegment
+from sqlfluff.core.rules.base import LintFix
+
+
+reflow_logger = logging.getLogger("sqlfluff.utils.reflow")
+
+
+@dataclass
+class _ReflowElement:
+    """Base reflow element class"""
+
+    segments: Sequence[RawSegment]
+    root_segment: BaseSegment
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return set(chain.from_iterable(seg.class_types for seg in self.segments))
+
+
+@dataclass
+class ReflowBlock(_ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single: the default (one single space)
+    # - close: no whitespace
+    before: str = "single"
+    after: str = "single"
+
+    def __post_init__(self):
+        """Infer spacing behaviour post-init."""
+        # NOTE: This should probably be configured differently in
+        # future, but this works initially. Delegating spacing
+        # configuration to the dialect or an explicit config
+        # are the two most likely options.
+        if {"start_bracket"} in self.class_types:
+            after = "close"
+        elif {"end_bracket"} in self.class_types:
+            before = "close"
+
+
+@dataclass
+class ReflowPoint(_ReflowElement):
+    """Class for keeping track of editable elements in reflow.
+
+    It holds segments which can be changed during a reflow operation
+    such as whitespace and newlines.
+
+    It holds no configuration and is influenced by the blocks either
+    side.
+    """
+
+    def trailing_whitespace_fixes(self) -> List[LintFix]:
+        """Fix any trailing whitespace detected.
+
+        Trailing whitespace is unique in that it can be detected
+        solely within the reflow point. It is any whitespace
+        preceeding a newline."""
+        # NOTE: We will need an end_of_file marker to do file ends.
+        reflow_logger.debug("TW: Checking %s", self)
+        if not self.class_types.intersection({"newline", "end_of_file"}):
+            return []
+        fixes = []
+        for idx, seg in enumerate(self.segments):
+            if (
+                seg.is_type("newline", "end_of_file")
+                and idx > 0
+                and self.segments[idx - 1].is_type("whitespace")
+            ):
```

Comment:
```suggestion
            if not (
                seg.is_type("newline", "end_of_file")
                and idx > 0
                and self.segments[idx - 1].is_type("whitespace")
            ):
                continue
```
Invert and save the indentation